doc/man-pages/man1/dpass.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.32
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "DPASS 1"
.TH DPASS 1 "2006-10-10" "OpenAFS" "AFS Command Reference"
.SH "NAME"
dpass \- Returns the DCE password for a new DCE account
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBdpass\fR [\fB\-cell\fR\ <\fIoriginal\ \s-1AFS\s0\ cell\ name\fR>] [\fB\-help\fR]
.PP
\&\fBdpass\fR [\fB\-c\fR\ <\fIoriginal\ \s-1AFS\s0\ cell\ name\fR>] [\fB\-h\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fBdpass\fR command returns the \s-1DCE\s0 password that an administrator
assigned to the issuer when using the \fBdm pass\fR command to migrate \s-1AFS\s0
user accounts into a \s-1DCE\s0 cell.
.PP
The \fBdpass\fR command, issued on an \s-1AFS\s0 client, requests the issuer's new
\&\s-1DCE\s0 password from the \s-1AFS\s0 cell specified with the \fB\-cell\fR argument.
.PP
The issuer must be authenticated as the \s-1AFS\s0 user whose \s-1AFS\s0 account was
moved into \s-1DCE\s0, and be able to provide the user's \s-1AFS\s0 password when
prompted by the \fBdpass\fR command.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-cell\fR <\fIcell name\fR>" 4
.IX Item "-cell <cell name>"
Specifies the name of the \s-1AFS\s0 cell from which the \s-1AFS\s0 account was moved
into \s-1DCE\s0 and from which to fetch the new \s-1DCE\s0 password.
.IP "\fB\-help\fR" 4
.IX Item "-help"
Prints the online help for this command. All other valid options are
ignored.
.SH "OUTPUT"
.IX Header "OUTPUT"
By default, the dpass command writes a message similar to the following to
the standard output stream.
.PP
.Vb 1
\&   Please read the following message before entering your password.
.Ve
.PP
.Vb 10
\&   This program will display your new, temporary DCE password on your
\&   terminal, and you should change the assigned password as soon as
\&   possible (from a DCE client).  The program assumes that the AFS cell
\&   uses the AFS Authentication Server and that an administrator used the
\&   utilities in the AFS/DFS Migration Toolkit to migrate the account from
\&   AFS to DCE. The password you enter should be the AFS password that was
\&   in effect when your DCE account was created; this is not necessarily
\&   the same password you have at the moment.  The cell name (which you may
\&   override with a command line option), must be the name of the AFS cell
\&   from which the authentication information was taken.
.Ve
.PP
To suppress this message, set the \s-1DPASS_NO_MESSAGE\s0 environment
variable. It is then possible to substitute a customized message if
desired by using a script similar to the following example:
.PP
.Vb 9
\&   #! /bin/csh
\&   echo "I<Start of customized message>"
\&   echo "I<Continuation of customized message>"
\&     .
\&     .
\&     .
\&   echo "I<Conclusion of customized message>"
\&   setenv DPASS_NO_MESSAGE
\&   dpass $*
.Ve
.PP
After the standard or customized message, if any, the dpass command
generates the following prompt for the original \s-1AFS\s0 password:
.PP
.Vb 2
\&   Original password for AFS cell <cell>:
\&   Re\-enter password to verify:
.Ve
.PP
If the \s-1AFS\s0 passwords match and are correct, the command reports the
temporary \s-1DCE\s0 password in the following message.
.PP
.Vb 1
\&   The new DCE password is: <Issuer's_temporary_DCE_password
.Ve
.SH "EXAMPLES"
.IX Header "EXAMPLES"
The following example returns the \s-1DCE\s0 password of the issuer, whose \s-1AFS\s0
account is in the \f(CW\*(C`abc.com\*(C'\fR cell. The \s-1DPASS_NO_MESSAGE\s0 variable has been
set to suppress the standard message.
.PP
.Vb 4
\&   % dpass
\&   Original password for AFS cell abc.com: <Issuer's_AFS_password>
\&   Re\-enter password to verify: <Issuer's_AFS_password>
\&   The new DCE password is: 8655\-\-eg8e\-dcdc\-8157
.Ve
.SH "PRIVILEGE REQUIRED"
.IX Header "PRIVILEGE REQUIRED"
The issuer must be authenticated as the \s-1AFS\s0 user for whom to display the
corresponding \s-1DCE\s0 password.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIdlog\fR\|(1)
.PP
\&\fBdm pass\fR reference page in \fI\s-1IBM\s0 \s-1AFS/DFS\s0 Migration Toolkit
Administration Guide and Reference\fR
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
\&\s-1IBM\s0 Corporation 2000. <http://www.ibm.com/> All Rights Reserved.
.PP
This documentation is covered by the \s-1IBM\s0 Public License Version 1.0.  It was
converted from \s-1HTML\s0 to \s-1POD\s0 by software written by Chas Williams and Russ
Allbery, based on work by Alf Wachsmann and Elizabeth Cassell.


NEWS
OpenAFS News -- history of user-visible changes.

* Changes incorporated in OpenAFS 1.4.0

** Addition of pthreading support in volserver.

** Threading optimization of RX libraries.

** Windows 95/98 client desupported.

** Digital UNIX/Tru64 desupported.

** Windows XP/2003 client includes persistent (across boots) cache support.

** Windows XP/2003 now supports a cache of up to 1.2gb.

** Windows XP/2003 has various improvements to the included GUI and command 
   line tools.

** Windows XP/2003 now supports AFS UNC paths.

** MacOS X (10.3) and Linux use system inodes instead of a private inode pool.

** Files larger than 2gb are supported in Unix clients, and in servers if the 
   --enable-largefile-fileserver switch is enabled at configure time.

** Obsolete and insecure auxiliary network daemons (rcp, rsh, ftpd et al)
   have been removed.

** Fileserver optimizations have been made to improve timeout and callback 
   breaking behavior to inaccessible clients or clients whose address has
   changed.

** Kerberos 5 support is integrated.

** -nosettime is now the default for afsd.  Use "-settime" to get the
   old behavior.

** Mountpoint directory information is now only faked for cross-cell
   mountpoints when using the -fakestat flag (e.g. for the directories
   under /afs, but not for most other volumes mounted inside the cell).
   The -fakestat-all switch can be used to fake information for all
   mountpoints.

** When fakestat is enabled on MacOSX, the Finder can be used to browse
   a fully-populated /afs directory.  However, this precludes reliable
   use of entire volumes as MacOS bundles (i.e. containing a Contents
   directory in the root of the volume).

** Mountpoint directory information can be faked by the cache manager,
   making operations such as stat'ing all cells under /afs much faster.
   This is enabled by passing -fakestat to afsd, but might not be stable
   on all platforms.

* Changes incorporated in OpenAFS 1.2.9

** The kaserver now defaults to not allowing interrealm authentication,
   due to security vulnerabilities in the krb4 protocol.  The new
   "-crossrealm" flag to the kaserver is provided to reenable interrealm
   authentication if desired.

** RedHat Linux 9.0 is now supported.

** Solaris 9 12/02 is now supported.  Solaris 7 and 8 x86 should now
   work again.

** On Linux machines using 2.2 series kernels, 2.2.19 or higher is now
   required.

** An OpenAFS 1.2.9 afsd will not work with kernel modules built from
   an earlier OpenAFS release.  In general, using a mismatched afsd and
   kernel modules set is unsupported; it is not recommended that you use
   such a configuration on a regular basis.

* Changes incorporated in OpenAFS 1.2.8

** Mountpoint directory information is now only faked for cross-cell
   mountpoits when using the -fakestat flag (e.g. for the directories
   under /afs, but not for most other volumes mounted inside the cell).
   The -fakestat-all switch can be used to fake information for all
   mountpoints.

** HPUX 11.0 is now supported.

** It is now possible for AFS to use Kerberos 5 directly, via rxkad 2b.
   See the OpenAFS 1.2.8 Release Notes for more information on using this
   capability.

** An NFS translator kernel module is now included and compiled by default
   for Solaris only.

* Changes incorporated in OpenAFS 1.2.7

** MacOS X 10.2 is now supported.  FreeBSD 4.3 and later support is included
   in this release, but is still under active development and should only
   be used by those doing active development on the OpenAFS FreeBSD client.

** When fakestat is enabled on MacOSX, the Finder can be used to browse
   a fully-populated /afs directory.  However, this precludes reliable
   use of entire volumes as MacOS bundles (i.e. containing a Contents
   directory in the root of the volume).

** The fileserver will now use Rx pings to determine if clients are reachable
   prior to allocating resources to them, to prevent asymmetric clients from
   consuming all fileserver resources.

* Changes incorporated in OpenAFS 1.2.6

** Mountpoint directory information can be faked by the cache manager,
   making operations such as stat'ing all cells under /afs much faster.
   This is enabled by passing -fakestat to afsd.

** Solaris 9 FCS and Solaris 7 and 8 x86 are now supported.

* Changes incorporated in OpenAFS 1.2.5

** A remote denial of service attack in the AIX and IRIX clients has
   been fixed.  Users of those platforms are strongly encouraged to
   upgrade.

** Fixed race conditions in fileserver that could result in crash.

* Changes incorporated in OpenAFS 1.2.4

** Server logfiles now more consistant about format in which hosts are
   referred to.

** vfsck on Solaris will now allow force runs (using -y flag) even if old 
   inodes exist.

* Changes incorporated in OpenAFS 1.2.3

** Cell aliases for dynroot can be specified in the CellAlias file in
   /usr/vice/etc or /usr/local/etc/openafs, in format "realname alias",
   one per line.  They can also be managed at runtime with "fs newalias"
   and "fs listaliases".

* Changes incorporated in OpenAFS 1.2.2

** Solaris 9 and Linux PA-RISC are now supported

** fileserver will not erroneously delay legitimate errors for 3 seconds
   after 10 errors are returned (e.g. stat() on a directory you can't read)

** Rx MTU calculation now works for Irix, Solaris and Linux

** If afsd is started with the -dynroot flag, /afs will be locally
   generated from the CellServDB.  AFSDB cells will be mounted
   automatically upon access.

** The namei fileserver allows vice "partitions" to be directories instead
   of partitions and will attach and display accordingly. Creating the file
   "AlwaysAttach" in the /vicepX directory is used as the trigger to attach it.

** TSM support for butc no longer requires editing a Makefile, simply
   specify the --enable-tivoli-tsm configure option.

** Linux builds no longer require source changes every time the kernel

src/dauth/dpass.c
/*
 * Copyright 2000, International Business Machines Corporation and others.
 * All Rights Reserved.
 * 
 * This software has been released under the terms of the IBM Public
 * License.  For details, see the LICENSE file in the top-level source
 * directory or online at http://www.openafs.org/dl/license10.html
 */

/*
 * dpass
 *
 * This program allows a user to discover the password generated for him
 * by the migration toolkit when migrating his password information
 * to the DCE.
 */

#include <afsconfig.h>
#include <afs/param.h>

RCSID
    ("$Header$");

#include <afs/stds.h>
#include <sys/types.h>
#include <rx/xdr.h>
#ifdef	AFS_AIX32_ENV
#include <signal.h>
#endif
#ifdef HAVE_STRING_H
#include <string.h>
#else
#ifdef HAVE_STRINGS_H
#include <strings.h>
#endif
#endif

#include <ubik.h>

#include <stdio.h>
#include <pwd.h>
#include <afs/com_err.h>
#include <afs/auth.h>
#include <afs/cellconfig.h>
#include <afs/afsutil.h>
#include <afs/cmd.h>
#include "adkint.h"
#include "assert.h"
#include <des.h>


char *msg[] = {
    "",
    "Please read the following message before entering your password.",
    "",
    "This program will display your new, temporary DCE password on your",
    "terminal, and you should change the assigned password as soon as",
    "possible (from a DCE client). The program assumes that your site uses",
    "the standard AFS authentication service provided by Transarc and that",
    "your initial account was created from the AFS authentication",
    "information by Transarc-supplied migration software.  If this is not",
    "the case, you should consult your system administrator.  The password",
    "you enter should be the AFS password that was in effect when your DCE",
    "account was created; this is not necessarily the same password you",
    "have at the moment. The cell name (which you may override with a",
    "command line option), must be the name of the AFS cell from which the",
    "authentication information was taken.",
    0
};

CommandProc(as, arock)
     char *arock;
     struct cmd_syndesc *as;
{
    int i;
    afs_int32 code;
    struct ktc_encryptionKey key;
    char cell[MAXKTCREALMLEN];
    char *cell_p;
    char prompt[1000];

    /*
     * We suppress the message, above, if this environment variable
     * is set. This allows the administrator to wrap dpass in a shell
     * script, if desired, in order to display more information
     * about registry conversion dates, etc.
     */
    if (!getenv("DPASS_NO_MESSAGE")) {
	for (i = 0; msg[i]; i++)
	    printf("%s\n", msg[i]);
    }

    if (as->parms[0].items) {
	struct afsconf_cell cellinfo;
	struct afsconf_dir *cdir;

	cell_p = as->parms[0].items->data;
	cdir = afsconf_Open(AFSDIR_CLIENT_ETC_DIRPATH);
	if (!cdir) {
	    fprintf(stderr,
		    "\nUnable to verify that \"%s\" is a valid cell name\nProceeding on the assumption that it is correct.\n",
		    cell_p);
	    exit(1);
	}
	code = afsconf_GetCellInfo(cdir, cell_p, 0, &cellinfo);
	if (code) {
	    fprintf(stderr,
		    "\nUnable to find information about cell \"%s\"\nProceeding on the assumption that this is a valid cell name.\n",
		    cell_p);
	} else {
	    strncpy(cell, cellinfo.name, sizeof(cell) - 1);
	    cell[sizeof(cell)-1] = '\0';
	    cell_p = cell;
	}
    } else {
	struct afsconf_dir *cdir;

	cdir = afsconf_Open(AFSDIR_CLIENT_ETC_DIRPATH);
	if (!cdir) {
	    fprintf(stderr,
		    "\nUnable to read the AFS client configuration file to get local cell name.\nTry specifying the cell with the -cell switch.\n");
	    exit(1);
	}
	afsconf_GetLocalCell(cdir, cell, sizeof(cell));
	cell_p = cell;
    }

    printf("\n");
    sprintf(prompt, "Original password for AFS cell %s: ", cell_p);
    code = ka_ReadPassword(prompt, 1, cell_p, &key);
    if (code) {
	fprintf(stderr, "Error reading password\n");
	exit(1);
    }
#define k(i) (key.data[i] & 0xff)
#define s(n) ((k(n) << 8) | k(n+1))
    printf("\nThe new DCE password is: %0.4x-%0.4x-%0.4x-%0.4x\n", s(0), s(2),
	   s(4), s(6));
}

main(argc, argv)
     int argc;
     char *argv[];
{
    struct cmd_syndesc *ts;
    afs_int32 code;
#ifdef	AFS_AIX32_ENV
    /*
     * The following signal action for AIX is necessary so that in case of a 
     * crash (i.e. core is generated) we can include the user's data section 
     * in the core dump. Unfortunately, by default, only a partial core is
     * generated which, in many cases, isn't too useful.
     */
    struct sigaction nsa;

    sigemptyset(&nsa.sa_mask);
    nsa.sa_handler = SIG_DFL;
    nsa.sa_flags = SA_FULLDUMP;
    sigaction(SIGSEGV, &nsa, NULL);
#endif
    ts = cmd_CreateSyntax(NULL, CommandProc, 0, "show new DCE passord");
    cmd_AddParm(ts, "-cell", CMD_SINGLE, CMD_OPTIONAL,
		"original AFS cell name");
    code = cmd_Dispatch(argc, argv);
    exit(code);
}
